### PR TITLE
CorfuStore: add Streaming API interfaces and stubs

### PR DESCRIPTION
**Pull request by hisundar**
## Overview
Introduces the streaming api for CorfuStore
These are just stubs for clients to code against.
Implementation will follow later.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        this.payload = payload;
+        this.metadata = metadata;
+        this.operation = operation;
+    }
+
+    /**
+     * Defines the type of the operation in this stream
+     */
+    public enum OperationType {
+        INSERT,
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: Create, update and delete.
We had a requirement to provide an explicit operation for 'create' and 'update'.

File: runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * Created by hisundar on 2019-10-18
+ */
+public interface Streamer {
+    /**
+     * A corfu update may have multiple updates belonging to different streams.
+     * This callback will return those updates as a list grouped by their Stream UUIDs.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    void onNext(Map<UUID, List<CorfuStreamEntry>> results);
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: We should hide the map data structure behind a result object of our own.
Let me know your thoughts.

File: runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     * Return a unique integer representing this streamer
+     * This is to allow CorfuStore may have to put these callback objects into a Set or Map.
+     */
+    int hashcode();
+
+    /**
+     * Return a human readable string representing this streamer.
+     * Log messages within CorfuStore will use this to identify different streamers.
+     */
+    String toString();
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: Should we rename this as `getName`?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 4]
```diff
@@ -151,4 +151,25 @@ public Query query(@Nonnull final String namespace) {
...
+    /**
+     * Subscribe to a specific a table in a namespace or the entire namespace.
+     * Objects returned will honor transactional boundaries
+     * @param streamer - callback context.
+     * @param namespace - the CorfuStore namespace to subscribe to.
+     * @param tableName - If specified, only updates from this table will be returned.
+     * @param timestamp - If specified, all stream updates from this timestamp will be returned
+     *                  - if null, only future updates will be returned.
+     */
+    public void subscribe(@Nonnull Streamer streamer, @Nonnull String namespace,
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: Not sure if it's a good idea to make the caller pass null.
Should we expose multiple signatures?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ *  NOTE: Ensure that the above protobuf generated classes are accessible to your JVM!
+ *
+ *  Created by hisundar on 2019-10-18
+ */
+public class CorfuStreamEntry<K extends Message, V extends Message, M extends Message> {
+    /**
+     * Key of the UFO stream entry
+     */
+    @Getter
+    K key;
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: Make all of these private and final.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 1 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java (hunk 2 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: good idea!

File: runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java (hunk 3 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: sure, toString() is the java convention right?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 4 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: without null, the number of signatures multiples quickly, that's why I just use null

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 5 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 4 above)
Comment (zalokhan) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: Ok.

File: runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package org.corfudb.runtime.collections;
+
+/**
+ * This is the callback interface that any client subscribe()ing to CorfuStore updates must implement.
+ *
+ * Created by hisundar on 2019-10-18
+ */
+public interface Streamer {
```
Comment (sneginhal) [inline runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: nit: StreamListener might be a better name than Streamer

Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/Streamer.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 7]
```diff
@@ -151,4 +151,25 @@ public Query query(@Nonnull final String namespace) {
...
                 namespace);
     }
+
+    /**
+     * Subscribe to a specific a table in a namespace or the entire namespace.
+     * Objects returned will honor transactional boundaries
+     * @param streamListener - callback context.
+     * @param namespace - the CorfuStore namespace to subscribe to.
+     * @param tableName - If specified, only updates from this table will be returned.
+     * @param timestamp - If specified, all stream updates from this timestamp will be returned
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: Did not understand the timestamp doc.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package org.corfudb.runtime.collections;
+
+import lombok.Getter;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * A wrapper around a map of stream UUID -> List of CorfuStreamEntry objects
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: Comment is unclear. What is the intention of this class and what is it going to be used for in expansion part.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
+package org.corfudb.runtime.collections;
+
+import com.google.protobuf.Message;
+import lombok.Getter;
+
+/**
+ * This will be the type of each entry returned by CorfuStore's Streamer Interface
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: I think you can reword the first sentence to:
Entry returned by CorfuStore's Streamer Interface.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * Metadata (ManagedResource) of the UFO stream entry
+     */
+    @Getter
+    private final M metadata;
+
+    /**
+     * Defines the type of the operation in this stream
+     */
+    public enum OperationType {
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: hopefully there are no migration / upgrade implications if a new operation is added later on.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package org.corfudb.runtime.collections;
+
+/**
+ * This is the callback interface that any client subscribe()ing to CorfuStore updates must implement.
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: remove the paranthesis in subscribing

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Created by hisundar on 2019-10-18
+ */
+public interface StreamListener {
+    /**
+     * A corfu update may have multiple updates belonging to different streams.
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: A corfu update can/may have multiple entries belonging to different streams.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    void onError(Throwable throwable);
+
+    /**
+     * Return if two streamer objects are equal.
+     * This is to allow CorfuStore may have to put these callback objects into a Set or Map.
+     */
+    boolean equals(StreamListener o);
+
+    /**
+     * Return a unique integer representing this streamer
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: replace: streamer/StreamListener.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+     * This is to allow CorfuStore may have to put these callback objects into a Set or Map.
+     */
+    int hashcode();
+
+    /**
+     * Return a human readable string representing this streamer.
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: Replace Streamer everywhere. [ Let us keep the terminology as lean and as close to code as possible.

Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java (hunk 13 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java (hunk 12 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java (hunk 11 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamListener.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 10 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: We should be able to add/remove from this Type as long as we can fix up the consumers.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 9 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java (hunk 8 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: It is basically a wrapper over MultiSMRObject where updates from different tables are returned, grouped by their stream UUIDs. Let me try to reword it better.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 7 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: The usage is 
1. take a snapshot which returns a timestamp
2. process all table entries at that timestamp
3. start streaming from that timestamp to process all remaining updates that happened after the snapshot was taken
perhaps I can put the above in the comment?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * A single corfu transaction can/may have updates to multiple tables.
+ * Each table in CorfuStore is represented by a stream identified by a UUID.
+ * This class returns those updates to the tables grouped by their stream UUIDs.
+ * This class allows further expansion of the returned values.
+ *
+ * created by hisundar on 2019-10-22
+ */
+public class CorfuStreamEntries {
```
Comment (sneginhal) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: Do we need CorfuStreamEntries if subscription is per table only?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    @Getter
+    private final Map<TableSchema, List<CorfuStreamEntry>> entries;
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: Shouldn't the key be just TableName - (namespace, tableName)?

File: runtime/src/main/java/org/corfudb/runtime/collections/TableSchema.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+@EqualsAndHashCode
+public class TableSchema<K extends Message, V extends Message, M extends Message> {
+    private final String tableName;
+    private final Class<K> keyClass;
+    private final Class<V> payloadClass;
+    private final Class<M> metadataClass;
+
+    public TableSchema(String tableName, Class<K> keyClass, Class<V> payloadClass, Class<M> metadataClass) {
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/TableSchema.java]: Can you add `@Nonnull` to the parameters except metadataClass which can be `@Nullable`?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        CREATE,
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: As discussed offline, can we remove the CREATE for the first phase.
Sorry for the confusion.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java (hunk 15 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: as discussed, we support both subscription per table and per selected tables in a namespace

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java (hunk 16 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntries.java]: callers would need the TableSchema anyway for processing the onNext() call, so I thought it just saves an extra parameter by just indexing on the TableSchema itself?

File: runtime/src/main/java/org/corfudb/runtime/collections/TableSchema.java (hunk 17 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/TableSchema.java]: done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 18 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: no problem. done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 7 above)
Comment (medhavidhawan) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: Yes please add it as comment.

**Pull request merged**